doc/source/concepts/concepts.rst

```diff
@@ -446,3 +447,133 @@ Below, we see an example of how to access the :code:`status` attribute of the re
+using :code:`apply_pip_deps` which is a callable that installs the required PyPI packages.
+
+Below is an example of using :code:`DepsPip` to specify a list of PyPI packages in an electron::
+
+    import covalent as ct
+    from covalent import DepsPip
+
+    @ct.electron(
+        deps_pip=DepsPip(packages=["numpy==0.23", "qiskit"]),
+    )
+    def task():
+    ...
+
+Alternatively, one can specify the path to a :code:`requirements.txt` file that contains the list of required packages::
+
+    @ct.electron(
+        deps_pip=DepsPip(reqs_path="/usr/foo/requirements.txt")
+    )
+
+~~~~~~~~
+DepsBash
+~~~~~~~~
+
+:code:`DepsBash` is the class that is responsible for managing the execution of bash commands that are required by an electron.
+
+:code:`__init_()`: The :code:`__init__()` constructor of :code:`DepsBash` accepts a list of bash commands as its argument.
+It overrides :code:`Deps`'s :code:`__init__()` by accepting :code:`apply_bash_commands` which is the callable that executes the commands and :code:`apply_args`
+which references the specified list of commands.
+
+:code:`apply_bash_commands`: This takes the list of commands and executes them as subprocesses in the same environment as the electron.
+
+Below is an example of using :code:`DepsBash` to specify a list of bash commands in an electron::
+
+    import covalent as ct
+    from covalent import DepsBash
+
+    @ct.electron(
+        deps_bash=DepsBash(["echo $PATH", "ssh foo@bar.com"]),
+    )
+    def task():
+    ...
+
+~~~~~~~~
+DepsCall
+~~~~~~~~
+
+:code:`DepsCall` is the class that is responsible for managing Python functions that need to be invoked in the same backend environment as the electron.
+
+:code:`__init__()`: :code:`DepsCall` :code:`__init__()` constructor takes in :code:`func` which is a callable
+that is invoked in the electron's environment. It also takes a list of :code:`args` and :code:`kwargs`
+which are passed as arguments when overriding the parent :code:`Deps` class.
+
+Below is an example of using :code:`DepsCall` to declare functions that are called before and after an electron is executed::
+
+    import covalent as ct
+    from covalent import DepsCall
+
+    def execute_before_electron():
+    ...
+
+    def shutdown_after_electron():
+    ...
+
+    @ct.electron(
+        call_before=DepsCall(execute_before_electron, args=[1, 2])
+        call_after=DepsCall(shutdown_after_electron),
+    )
+    def task():
+    ...
+
+
+*Note*: It's also possible to implicitly declare multiple kinds of dependencies in an electron::
+
+    import covalent as ct
+    from covalent import DepsPip, DepsBash, DepsCall
+
+    def execute_before_electron(a, b):
+    ...
+
+    def shutdown_after_electron():
+    ...
+
+    @ct.electron(
+        deps_pip=DepsPip(packages=["numpy==0.23", "qiskit"]),
+        deps_bash=DepsBash(commands=["echo $PATH", "ssh foo@bar.com"]),
+        call_before=DepsCall(execute_before_electron, args=(1, 2)),
```

Comment:
Hey @Emmanuel289  I dont know if this got updated on your end, but seems like `DepsCall` is the parent class for all Deps, which means that any Deps object can be passed to `call_before` and `call_after` even `DepsBash` etc..